test/hooks.on-listen.test.js

```diff
@@ -0,0 +1,614 @@
+'use strict'
+
+const { t, before } = require('tap')
+const Fastify = require('../fastify')
+const fp = require('fastify-plugin')
+const dns = require('dns').promises
+
+let localhost
+
+before(async function () {
+  const lookup = await dns.lookup('localhost')
+  localhost = lookup.address
+})
+
+t.test('localhost onListen should be called in order', t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 1, '2nd called in root')
+    done()
+  })
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+t.test('localhost async onListen should be called in order', async t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, '1st async called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 1, '2nd async called in root')
+  })
+
+  await fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+t.test('localhost onListen should manage error in sync', t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.pass('called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st sync called in root')
+    done(new Error('FAIL ON LISTEN'))
+  })
+
+  fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
+t.test('localhost onListen should manage error in async', async t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.pass('called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, '1st async called in root')
+    throw new Error('FAIL ON LISTEN')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.pass('called in root')
+  })
+
+  await fastify.listen({
+    host: 'localhost',
+    port: 0
+  })
+})
```

Comment:
this test seems wrong in many ways.

t.plan is 2, but we have three calls. also the first addHook onListen  doesnt increment order, lastly we just pass  in the last hook handler, despite failing the unit test.

Did you mean:

```suggestion
t.test('localhost onListen should manage error in async', async t => {
  t.plan(1)
  const fastify = Fastify()
  t.teardown(fastify.close.bind(fastify))
  let order = 0

  fastify.addHook('onListen', async function () {
    t.equal(order++, 0, '1st async called in root')
    throw new Error('FAIL ON LISTEN')
  })

  fastify.addHook('onListen', async function () {
    t.fail('should not been called')
  })

  await fastify.listen({
    host: 'localhost',
    port: 0
  })
})
```

Reply:
Yes, this is far more appropriate and functional than what is currently implemented. We need to redo a lot of the testing (we have been learning as we go). Sorry about that.

Reply:
Nothing to be sorry about.